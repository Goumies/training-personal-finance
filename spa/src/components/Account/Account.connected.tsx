import React from "react";
import { StateContext, State } from "../../StateContext/StateContext";
import { AccountComponent, OwnProps } from "./Account";

export const Account: React.StatelessComponent<OwnProps> = () => (
  <StateContext.Consumer>{(state: State) => <AccountComponent operations={state.operations} />}</StateContext.Consumer>
);
