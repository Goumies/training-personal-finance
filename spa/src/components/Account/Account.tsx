import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { OperationHistory } from "../OperationHistory";
import { Balance } from "../Balance";
import { AddOperation } from "../AddOperation";

export interface Props {
  operations: IOperation[];
}

type State = Props;

export class Account extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      operations: props.operations
    };
  }

  private onNewOperation = (operation: IOperation): void => {
    this.setState({
      ...this.state,
      operations: [...this.state.operations, operation]
    });
  };

  private onOperationDeletion = (id: string): void => {
    const operations: IOperation[] = [...this.state.operations];
    const index: number = operations.findIndex(({ id: currentId }: IOperation) => currentId === id);
    if (index !== -1) {
      operations.splice(index, 1);
      this.setState({
        ...this.state,
        operations
      });
    }
  };

  render() {
    const balance: number = this.state.operations.map(extractAmount).reduce(sum, 0);
    return (
      <div>
        <AddOperation onNewOperation={this.onNewOperation} />
        <OperationHistory operations={this.state.operations} onDelete={this.onOperationDeletion} />
        <Balance amount={balance} />
      </div>
    );
  }
}

function extractAmount({ amount }: IOperation): number {
  return amount;
}

function sum(accumulated: number, current: number): number {
  return accumulated + current;
}
