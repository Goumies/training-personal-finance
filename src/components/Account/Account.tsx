import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { OperationHistory } from "../OperationHistory";
import { Balance } from "../Balance";

interface Props {
  operations: IOperation[];
}

export const Account: React.StatelessComponent<Props> = ({ operations }: Props) => {
  const balance: number = operations.map(extractAmount).reduce(sum, 0);
  return (
    <div>
      <OperationHistory operations={operations} />
      <Balance amount={balance} />
    </div>
  );
};

function extractAmount({ amount }: IOperation): number {
  return amount;
}

function sum(accumulated: number, current: number): number {
  return accumulated + current;
}
