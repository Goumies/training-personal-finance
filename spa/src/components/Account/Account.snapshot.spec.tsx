import React from "react";
import renderer from "react-test-renderer";
import { Account, Props } from "./Account";

it("renders correctly", () => {
  const props: Props = {
    operations: [
      {
        id: "0",
        date: new Date(2019, 4, 24),
        amount: 10
      },
      {
        id: "1",
        date: new Date(2019, 4, 25),
        amount: -5
      }
    ]
  };
  const tree = renderer.create(<Account {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
