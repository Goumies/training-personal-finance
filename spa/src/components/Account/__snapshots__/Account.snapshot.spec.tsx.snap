// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="add-operation"
  >
    <input
      className="input-amount"
      data-e2e="account-new-operation-input-amount"
      onChange={[Function]}
      onFocus={[Function]}
      type="number"
      value=""
    />
    <button
      className="add-amount mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent"
      data-e2e="account-new-operation-add-amount"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <div>
    <div
      className="operation"
      data-e2e="account-operation"
    >
      <div
        className="labeled-debit-or-credit"
      >
        <div
          className="label-container"
        >
          <span
            className="date"
          >
            2019-5-24
          </span>
        </div>
        <div
          className="amount-container"
        >
          <div
            className="debit-or-credit"
          >
            <span
              className="credit"
            >
              <span
                className="value"
              >
                +10
              </span>
            </span>
            <span
              className="debit"
            />
          </div>
        </div>
      </div>
      <div
        className="actions-container"
      >
        <div
          className="actions-content"
        >
          <div
            className="actions"
          >
            <button
              className="delete mdl-button mdl-js-button mdl-button--icon"
              data-e2e="account-operation-delete"
              onClick={[Function]}
            >
              <i
                className="material-icons"
              >
                delete
              </i>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="operation"
      data-e2e="account-operation"
    >
      <div
        className="labeled-debit-or-credit"
      >
        <div
          className="label-container"
        >
          <span
            className="date"
          >
            2019-5-25
          </span>
        </div>
        <div
          className="amount-container"
        >
          <div
            className="debit-or-credit"
          >
            <span
              className="credit"
            />
            <span
              className="debit"
            >
              <span
                className="value"
              >
                -5
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        className="actions-container"
      >
        <div
          className="actions-content"
        >
          <div
            className="actions"
          >
            <button
              className="delete mdl-button mdl-js-button mdl-button--icon"
              data-e2e="account-operation-delete"
              onClick={[Function]}
            >
              <i
                className="material-icons"
              >
                delete
              </i>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="balance"
    data-e2e="account-balance"
  >
    <div
      className="labeled-debit-or-credit"
    >
      <div
        className="label-container"
      >
        <span>
          Balance
        </span>
      </div>
      <div
        className="amount-container"
      >
        <div
          className="debit-or-credit"
        >
          <span
            className="credit"
          >
            <span
              className="value"
            >
              +5
            </span>
          </span>
          <span
            className="debit"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
